packages/catalog-model/src/entity/ref.ts

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { CompoundEntityRef, EntityName, EntityRef } from '../types';
+
+/**
+ * The context of defaults that entity reference parsing happens within.
+ */
+export type EntityRefContext = {
+  /** The default kind, if none is given in the reference */
+  defaultKind?: string;
+  /** The default namespace, if none is given in the reference */
+  defaultNamespace?: string;
+};
+
+/**
+ * Parses an entity reference, either on string or compound form, and always
+ * returns a complete entity name including kind, namespace and name.
+ *
+ * This function automatically assumes the default namespace "default" unless
+ * otherwise specified as part of the options, and will throw an error if no
+ * kind was specified in the input reference and no default kind was given.
+ *
+ * @param ref The reference to parse
+ * @param context The context of defaults that the parsing happens within
+ * @returns A complete entity name
+ */
+export function parseEntityName(
+  ref: EntityRef,
+  context: EntityRefContext = {},
+): EntityName {
+  const { kind, namespace, name } = parseEntityRef(ref, {
+    defaultNamespace: 'default',
+    ...context,
+  });
+
+  if (!kind) {
+    throw new Error(
+      `Entity reference ${namespace}/${name} did not contain a kind`,
+    );
+  }
+
+  // This is a corner case - when the ref contained a namespace but it was the
+  // empty string (so the default did not kick in)
+  if (!namespace) {
+    throw new Error(
+      `Entity reference ${kind}:${name} did not contain a namespace`,
+    );
+  }
+
+  return {
+    kind,
+    namespace: namespace!,
+    name,
+  };
+}
+
+/**
+ * Parses an entity reference, either on string or compound form, and returns
+ * a structure with a name, and optional kind and namespace.
+ *
+ * The options object can contain default values for the kind and namespace,
+ * that will be used if the input reference did not specify any.
+ *
+ * @param ref The reference to parse
+ * @param context The context of defaults that the parsing happens within
+ * @returns The compound form of the reference
+ */
+export function parseEntityRef(
```

Comment:
Was the intention of this function to have a separate function that only does the parsing bit, where applying defaults happen separately? Right now I don't see why we need both of them.

Either way I'm thinking we only export one function that does the full transform, and then possibly add the intermediate one if needed later.

Reply:
Yes, it is useful in the cases where you do want to know what the actual details of the input was. But yeah if you think that we should hide it from the exports then sure.

Reply:
Then it would be better to not have the defaulting mechanic in the parsing function. But yes, I think we should stick with only exporting a single function anyway.